spec/Section 3 -- Type System.md

```diff
@@ -775,13 +775,26 @@ An input object is never a valid result.
 
 **Input Coercion**
 
-The input to an input object should be an unordered map, otherwise an error
-should be thrown. The result of the coercion is an unordered map, with an
-entry for each input field, whose key is the name of the input field.
-The value of an entry in the coerced map is the result of input coercing the
-value of the entry in the input with the same key; if the input does not have a
-corresponding entry, the value is the result of coercing null. The input
-coercion above should be performed according to the input coercion rules of the
+The value for an input object should be an input object literal or an unordered
+map, otherwise an error should be thrown. This unordered map should not contain
+any entries with names not defined by a field of this input object type,
+otherwise an error should be thrown.
+
+If any non-nullable fields defined by the input object do not have corresponding
+entries in the original value, were provided a variable for which a value was
+not provided, or for which the value {null} was provided, an error should
+be thrown.
```

Comment:
With the introduction of a null literal, do we need a separate concept of "non-nullable" and "required"? For example, should it be possible to declare an input object field which must be provided, but its value may be null?

    { required_field: 42 } # valid
    { required_field: null } # valid
    { } # invalid
    { required_field: $var } # valid if $var is required but $var may be nullable...I guess?